src/main/java/com/minecolonies/coremod/research/GlobalResearch.java

```diff
@@ -256,14 +438,396 @@ public void setParent(final String id)
+    /**
+     * Gets the branch for a research from a JSON object, if it exists and is valid, or "parseerrors" otherwise.
+     *
+     * @param researchJson        A json object to retrieve the ID from.
+     * @param resourceLocation    The {@link ResourceLocation} of the json being parsed.
+     * @return                    The Research Branch as a String.
+     */
+    private String getBranch(final JsonObject researchJson, final ResourceLocation resourceLocation)
+    {
+        if (researchJson.has(RESEARCH_BRANCH_PROP) && researchJson.get(RESEARCH_BRANCH_PROP).isJsonPrimitive() && researchJson.get(RESEARCH_BRANCH_PROP).getAsJsonPrimitive().isString())
+        {
+            return researchJson.get(RESEARCH_BRANCH_PROP).getAsString();
+        }
+        else
+        {
+            Log.getLogger().error("Error in Research Branch for" + resourceLocation);
+            return "parseerrors";
+        }
+    }
+
+    /**
+     * Gets the required university level from a JSON object, if it exists and is valid, or returns 1 if not.
+     *
+     * @param researchJson        A json object to retrieve the requiredUniversityLevel from.
+     * @return                    The required university level as an integer.
+     */
+    private int getUniversityLevel(final JsonObject researchJson)
+    {
+        if (researchJson.has(RESEARCH_UNIVERSITY_LEVEL_PROP) && researchJson.get(RESEARCH_UNIVERSITY_LEVEL_PROP).isJsonPrimitive() && researchJson.get(RESEARCH_UNIVERSITY_LEVEL_PROP).getAsJsonPrimitive().isNumber())
+        {
+            return researchJson.get(RESEARCH_UNIVERSITY_LEVEL_PROP).getAsNumber().intValue();
+        }
+        else
+        {
+            Log.getLogger().info("No declared university level for " + this.branch + "/" + this.id );
+            return 1;
+        }
+    }
+
+    /**
+     * Gets the optional icon location from a research json, if present.  If not available, or if requests a file or block that does not exist, returns an empty string.
+     * @param researchJson        A json object to retrieve the requiredUniversityLevel from.
+     * @return                    The required university level as an integer.
+     */
+    private String getResearchIcon(final JsonObject researchJson)
+    {
+        if (researchJson.has(RESEARCH_ICON_PROP) && researchJson.get(RESEARCH_ICON_PROP).isJsonPrimitive() && researchJson.get(RESEARCH_ICON_PROP).getAsJsonPrimitive().isString())
+        {
+            final String[] iconParts = researchJson.get(RESEARCH_ICON_PROP).getAsString().split(":");
+            final String[] outputString = new String[3];
+            // Do preliminary validation here, as later uses will always be in UI space.
+            if(iconParts.length > 3)
+            {
+                Log.getLogger().info("Malformed icon property for " + this.branch + "/" + this.id + ".  Icons may contain at most namespace:identifier:count.");
+                return "";
+            }
+
+            if(iconParts.length == 3)
```

Comment:
This seems overly complex to just get a resource location?

Reply:
Needs to handle and distinguish textures or items, with and without namespaces, and for items with or without counts, which may be entered by typo if users and not just developers.

I can move the validation stuff elsewhere, but most of it is important: a bad texture location just spams the console each time it’s called , but an unparsable number would just crash Minecraft entirely with an awful stack trace since it’s going to GUI space.

Reply:
ok